File: src/cmd/flux-queue.c
/************************************************************\
 * Copyright 2020 Lawrence Livermore National Security, LLC
 * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
 *
 * This file is part of the Flux resource manager framework.
 * For details, see https://github.com/flux-framework.
 *
 * SPDX-License-Identifier: LGPL-3.0
\************************************************************/

/* flux-queue.c - control job manager queue
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <assert.h>
#include <jansson.h>
#include <argz.h>
#include <flux/core.h>
#include <flux/optparse.h>

#include "src/common/libutil/log.h"

int cmd_enable (optparse_t *p, int argc, char **argv);
int cmd_disable (optparse_t *p, int argc, char **argv);
int cmd_start (optparse_t *p, int argc, char **argv);
int cmd_stop (optparse_t *p, int argc, char **argv);
int cmd_status (optparse_t *p, int argc, char **argv);
int cmd_drain (optparse_t *p, int argc, char **argv);
int cmd_idle (optparse_t *p, int argc, char **argv);

int stdin_flags;

static struct optparse_option global_opts[] =  {
    OPTPARSE_TABLE_END
};

static struct optparse_option stop_opts[] = {
    { .name = "verbose", .key = 'v',
      .usage = "Display more detail about internal job manager state",
    },
    { .name = "quiet", .has_arg = 0,
      .usage = "Display only errors",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option start_opts[] = {
    { .name = "verbose", .key = 'v',
      .usage = "Display more detail about internal job manager state",
    },
    { .name = "quiet", .has_arg = 0,
      .usage = "Display only errors",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option status_opts[] = {
    { .name = "verbose", .key = 'v',
      .usage = "Display more detail about internal job manager state",
    },
    { .name = "queue", .key = 'q', .has_arg = 1, .arginfo = "NAME",
      .usage = "Specify queue to show (default all)",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option enable_opts[] = {
    { .name = "queue", .key = 'q', .has_arg = 1, .arginfo = "NAME",
      .usage = "Specify queue to enable",
    },
    { .name = "all", .key = 'a', .has_arg = 0,
      .usage = "Force command to apply to all queues if none specified",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option disable_opts[] = {
    { .name = "queue", .key = 'q', .has_arg = 1, .arginfo = "NAME",
      .usage = "Specify queue to disable",
    },
    { .name = "all", .key = 'a', .has_arg = 0,
      .usage = "Force command to apply to all queues if none specified",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option drain_opts[] =  {
    { .name = "timeout", .key = 't', .has_arg = 1, .arginfo = "DURATION",
      .usage = "timeout after DURATION",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_option idle_opts[] =  {
    { .name = "timeout", .key = 't', .has_arg = 1, .arginfo = "DURATION",
      .usage = "timeout after DURATION",
    },
    { .name = "quiet", .has_arg = 0,
      .usage = "Only display pending job count if nonzero",
    },
    OPTPARSE_TABLE_END
};

static struct optparse_subcommand subcommands[] = {
    { "enable",
      "[OPTIONS]",
      "Enable job submission",
      cmd_enable,
      0,
      enable_opts,
    },
    { "disable",
      "[OPTIONS] [message ...]",
      "Disable job submission",
      cmd_disable,
      0,
      disable_opts,
    },
    { "start",
      "[OPTIONS]",
      "Start scheduling",
      cmd_start,
      0,
      start_opts,
    },
    { "stop",
      "[OPTIONS] [message ...]",
      "Stop scheduling",
      cmd_stop,
      0,
      stop_opts,
    },
    { "status",
      "[OPTIONS]",
      "Get queue status",
      cmd_status,
      0,
      status_opts,
    },
    { "drain",
      "[OPTIONS]",
      "Wait for queue to become empty.",
      cmd_drain,
      0,
      drain_opts
    },
    { "idle",
      "[OPTIONS]",
      "Wait for queue to become idle.",
      cmd_idle,
      0,
      idle_opts
    },
    OPTPARSE_SUBCMD_END
};

int usage (optparse_t *p, struct optparse_option *o, const char *optarg)
{
    struct optparse_subcommand *s;
    optparse_print_usage (p);
    fprintf (stderr, "\n");
    fprintf (stderr, "Common commands from flux-queue:\n");
    s = subcommands;
    while (s->name) {
        fprintf (stderr, "   %-15s %s\n", s->name, s->doc);
        s++;
    }
    exit (1);
}

int main (int argc, char *argv[])
{
    char *cmdusage = "[OPTIONS] COMMAND ARGS";
    optparse_t *p;
    int optindex;
    int exitval;

    log_init ("flux-queue");

    p = optparse_create ("flux-queue");

    if (optparse_add_option_table (p, global_opts) != OPTPARSE_SUCCESS)
        log_msg_exit ("optparse_add_option_table() failed");

    /* Override help option for our own */
    if (optparse_set (p, OPTPARSE_USAGE, cmdusage) != OPTPARSE_SUCCESS)
        log_msg_exit ("optparse_set (USAGE)");

    /* Override --help callback in favor of our own above */
    if (optparse_set (p, OPTPARSE_OPTION_CB, "help", usage) != OPTPARSE_SUCCESS)
        log_msg_exit ("optparse_set() failed");

    /* Don't print internal subcommands, we do it ourselves */
    if (optparse_set (p, OPTPARSE_PRINT_SUBCMDS, 0) != OPTPARSE_SUCCESS)
        log_msg_exit ("optparse_set (PRINT_SUBCMDS)");

    if (optparse_reg_subcommands (p, subcommands) != OPTPARSE_SUCCESS)
        log_msg_exit ("optparse_reg_subcommands");

    if ((optindex = optparse_parse_args (p, argc, argv)) < 0)
        exit (1);

    if ((argc - optindex == 0)
        || !optparse_get_subcommand (p, argv[optindex])) {
        usage (p, NULL, NULL);
        exit (1);
    }

    if ((exitval = optparse_run_subcommand (p, argc, argv)) < 0)
        exit (1);

    optparse_destroy (p);
    log_fini ();
    return (exitval);
}

/* Parse free arguments into a space-delimited message.
 * On error, exit complaning about parsing 'name'.
 * Caller must free the resulting string
 */
static char *parse_arg_message (char **argv, const char *name)
{
    char *argz = NULL;
    size_t argz_len = 0;
    error_t e;

    if ((e = argz_create (argv, &argz, &argz_len)) != 0)
        log_errn_exit (e, "error parsing %s", name);
    argz_stringify (argz, argz_len, ' ');
    return argz;
}

void alloc_admin (flux_t *h,
                  bool verbose,
                  bool quiet,
                  int query_only,
                  int start,
                  const char *reason)
{
    flux_future_t *f;
    int free_pending;
    int alloc_pending;
    int queue_length;
    int running;

    if (!(f = flux_rpc_pack (h,
                             "job-manager.alloc-admin",
                             FLUX_NODEID_ANY,
                             0,
                             "{s:b s:b s:s}",
                             "query_only",
                             query_only,
                             "start",
                             start,
                             "reason",
                             reason ? reason : "")))
        log_err_exit ("error sending alloc-admin request");
    if (flux_rpc_get_unpack (f,
                             "{s:b s:s s:i s:i s:i s:i}",
                             "start",
                             &start,
                             "reason",
                             &reason,
                             "queue_length",
                             &queue_length,
                             "alloc_pending",
                             &alloc_pending,
                             "free_pending",
                             &free_pending,
                             "running",
                             &running) < 0)
        log_msg_exit ("alloc-admin: %s", future_strerror (f, errno));
    if (!quiet) {
        printf ("Scheduling is %s%s%s\n",
                start ? "started" : "stopped",
                reason && strlen (reason) > 0 ? ": " : "",
                reason ? reason : "");
    }
    if (verbose) {
        printf ("%d alloc requests queued\n", queue_length);
        printf ("%d alloc requests pending to scheduler\n", alloc_pending);
        printf ("%d free requests pending to scheduler\n", free_pending);
        printf ("%d running jobs\n", running);
    }
    flux_future_destroy (f);
}

static void add_string_if_set (json_t *o, const char *key, const char *val)
{
    if (val) {
        json_t *str = json_string (val);
        if (!str || json_object_set_new (o, key, str) < 0) {
            json_decref (str);
            log_msg_exit ("out of memory");
        }
    }
}

static void queue_enable (flux_t *h,
                          const char *name,
                          bool enable,
                          const char *reason,
                          bool all)
{
    json_t *payload;
    flux_future_t *f;

    if (!(payload = json_pack ("{s:b s:b}",
                               "enable", enable ? 1 : 0,
                               "all", all ? 1 : 0)))
        log_msg_exit ("out of memory");
    add_string_if_set (payload, "name", name);
    add_string_if_set (payload, "reason", reason);
    f = flux_rpc_pack (h, "job-manager.queue-enable", 0, 0, "O", payload);
    if (!f || flux_rpc_get (f, NULL) < 0)
        log_msg_exit ("%s", future_strerror (f, errno));
    flux_future_destroy (f);
    json_decref (payload);
}

typedef void (*queue_status_output_f) (const char *name,
                                       bool enable,
                                       const char *reason);

static void queue_status_one (flux_t *h,
                              const char *name,
                              queue_status_output_f out_cb)
{
    json_t *payload;
    flux_future_t *f;
    int enable;
    const char *reason = NULL;

    if (!(payload = json_object ()))
        log_msg_exit ("out of memory");
    add_string_if_set (payload, "name", name);
    f = flux_rpc_pack (h, "job-manager.queue-status", 0, 0, "O", payload);
    if (!f || flux_rpc_get_unpack (f,
                                   "{s:b s?s}",
                                   "enable", &enable,
                                   "reason", &reason))
        log_msg_exit ("%s", future_strerror (f, errno));
    out_cb (name, enable, reason);
    flux_future_destroy (f);
    json_decref (payload);
}

static void queue_status (flux_t *h,
                          const char *name,
                          queue_status_output_f out_cb)
{
    if (!name) {
        json_t *queues;
        flux_future_t *f;

        f = flux_rpc (h, "job-manager.queue-list", NULL, 0, 0);
        if (!f || flux_rpc_get_unpack (f,
                                       "{s:o}",
                                       "queues", &queues))
            log_msg_exit ("%s", future_strerror (f, errno));
        if (json_array_size (queues) > 0) {
            size_t index;
            json_t *value;
            json_array_foreach (queues, index, value) {
                queue_status_one (h, json_string_value (value), out_cb);
            }
        }
        else
            queue_status_one (h, NULL, out_cb);
        flux_future_destroy (f);
    }
    else
        queue_status_one (h, name, out_cb);
}

int cmd_enable (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    const char *name = optparse_get_str (p, "queue", NULL);
    bool all = optparse_hasopt (p, "all");

    if (argc - optindex > 0) {
        optparse_print_usage (p);
        exit (1);
    }
    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    queue_enable (h, name, true, NULL, all);
    flux_close (h);
    return (0);
}

int cmd_disable (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    const char *name = optparse_get_str (p, "queue", NULL);
    bool all = optparse_hasopt (p, "all");
    char *reason = NULL;

    if (argc - optindex > 0)
        reason = parse_arg_message (argv + optindex, "reason");
    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    queue_enable (h, name, false, reason, all);
    flux_close (h);
    free (reason);
    return (0);
}

int cmd_start (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);

    if (argc - optindex > 0) {
        optparse_print_usage (p);
        exit (1);
    }
    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    alloc_admin (h,
                 optparse_hasopt (p, "verbose"),
                 optparse_hasopt (p, "quiet"),
                 0,
                 1,
                 NULL);
    flux_close (h);
    return (0);
}

int cmd_stop (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    char *reason = NULL;

    if (argc - optindex > 0)
        reason = parse_arg_message (argv + optindex, "reason");
    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    alloc_admin (h,
                 optparse_hasopt (p, "verbose"),
                 optparse_hasopt (p, "quiet"),
                 0,
                 0,
                 reason);
    flux_close (h);
    free (reason);
    return (0);
}

static void print_enable_status (const char *name,
                                 bool enable,
                                 const char *reason)
{
    if (enable) {
        printf ("%s%sJob submission is enabled\n",
                name ? name : "",
                name ? ": " : "");
    }
    else {
        printf ("%s%sJob submission is disabled: %s\n",
                name ? name : "",
                name ? ": " : "",
                reason);
    }
}

int cmd_status (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    const char *name = optparse_get_str (p, "queue", NULL);

    if (argc - optindex > 0) {
        optparse_print_usage (p);
        exit (1);
    }
    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    queue_status (h, name, print_enable_status);

    alloc_admin (h,
                 optparse_hasopt (p, "verbose"),
                 false,
                 1,
                 0,
                 NULL);
    flux_close (h);
    return (0);
}

int cmd_drain (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    double timeout = optparse_get_duration (p, "timeout", -1.);
    flux_future_t *f;

    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    if (argc - optindex != 0) {
        optparse_print_usage (p);
        exit (1);
    }
    if (!(f = flux_rpc (h, "job-manager.drain", NULL, FLUX_NODEID_ANY, 0)))
        log_err_exit ("flux_rpc");
    if (flux_future_wait_for (f, timeout) < 0 || flux_rpc_get (f, NULL) < 0)
        log_msg_exit ("drain: %s", errno == ETIMEDOUT
                                   ? "timeout" : future_strerror (f, errno));
    flux_future_destroy (f);
    flux_close (h);
    return (0);
}

int cmd_idle (optparse_t *p, int argc, char **argv)
{
    flux_t *h;
    int optindex = optparse_option_index (p);
    double timeout = optparse_get_duration (p, "timeout", -1.);
    flux_future_t *f;
    int pending;

    if (!(h = flux_open (NULL, 0)))
        log_err_exit ("flux_open");
    if (argc - optindex != 0) {
        optparse_print_usage (p);
        exit (1);
    }
    if (!(f = flux_rpc (h, "job-manager.idle", NULL, FLUX_NODEID_ANY, 0)))
        log_err_exit ("flux_rpc");
    if (flux_future_wait_for (f, timeout) < 0
            || flux_rpc_get_unpack (f, "{s:i}", "pending", &pending) < 0)
        log_msg_exit ("idle: %s", errno == ETIMEDOUT
                                   ? "timeout" : future_strerror (f, errno));
    if (!optparse_hasopt (p, "quiet") || pending > 0)
        printf ("%d pending jobs\n", pending);
    flux_future_destroy (f);
    flux_close (h);
    return (0);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */

File: src/cmd/flux-queue.py
##############################################################
# Copyright 2022 Lawrence Livermore National Security, LLC
# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
#
# This file is part of the Flux resource manager framework.
# For details, see https://github.com/flux-framework.
#
# SPDX-License-Identifier: LGPL-3.0
##############################################################

import argparse
import errno
import logging
import sys

import flux
from flux.queue import QueueList
from flux.util import AltField, FilterActionSetUpdate, UtilConfig, parse_fsd


def print_enable_status(name, status):
    if name:
        print(f"{name}: ", end="")
    if status["enable"]:
        print("Job submission is enabled")
    else:
        disable_reason = status["disable_reason"]
        print(f"Job submission is disabled: {disable_reason}")


def print_start_status(name, status):
    if name:
        print(f"{name}: ", end="")
    if status["start"]:
        print("Scheduling is started")
    else:
        reason = ": " + status["stop_reason"] if "stop_reason" in status else None
        print(f"""Scheduling is stopped{reason if reason else ""}""")


def print_queue_status(name, status):
    print_enable_status(name, status)
    print_start_status(name, status)


def queue_status_get_one(handle, name=None):
    payload = {"name": name} if name else {}
    future = handle.rpc("job-manager.queue-status", payload, nodeid=0)
    try:
        status = future.get()
    except EnvironmentError:
        LOGGER.error("status: {0}".format(future.error_string()))
        sys.exit(1)
    return status


def queue_status_get(handle, queue, verbose):
    statuses = []
    future = handle.rpc("job-manager.queue-list")
    try:
        qlist = future.get()
    except EnvironmentError:
        LOGGER.error("list: {0}".format(future.error_string()))
        sys.exit(1)
    if not queue or verbose:
        if len(qlist["queues"]) > 0:
            for q in qlist["queues"]:
                status = queue_status_get_one(handle, q)
                statuses.append((q, status))
        else:
            status = queue_status_get_one(handle)
            statuses.append((queue, status))
    else:
        status = queue_status_get_one(handle, queue)
        statuses.append((queue, status))
    return statuses


def queue_status(handle, queue, verbose, out_cb):
    statuses = queue_status_get(handle, queue, verbose)
    for name, status in statuses:
        out_cb(name, status)


def alloc_query(handle):
    future = handle.rpc("job-manager.alloc-query")
    try:
        query = future.get()
    except EnvironmentError:
        LOGGER.error("alloc-query: {0}".format(future.error_string()))
        sys.exit(1)
    print("{0} alloc requests queued".format(query["queue_length"]))
    print("{0} alloc requests pending to scheduler".format(query["alloc_pending"]))
    print("{0} running jobs".format(query["running"]))


def status(args):
    handle = flux.Flux()
    queue_status(handle, args.queue, args.verbose, print_queue_status)
    if args.verbose:
        alloc_query(handle)


class FluxQueueConfig(UtilConfig):
    """flux-queue specific user configuration class"""

    builtin_formats = {}
    builtin_formats["list"] = {
        "default": {
            "description": "Default flux-queue list format string",
            "format": (
                "?:{queuem:<8.8} "
                "{color_enabled}{enabled:>2}{color_off} "
                "{color_started}{started:>2}{color_off} "
                "{defaults.timelimit!F:>8} "
                "{limits.timelimit!F:>8} "
                "{limits.range.effective.nnodes:>10} "
                "{limits.range.effective.ncores:>10} "
                "{limits.range.effective.ngpus:>10}"
            ),
        },
        "policy": {
            "description": "flux-queue list with policy limits only",
            "format": (
                "?:{queuem:<8.8} "
                "{color_enabled}{enabled:>2}{color_off} "
                "{color_started}{started:>2}{color_off} "
                "{defaults.timelimit!F:>8} "
                "{limits.timelimit!F:>8} "
                "{limits.range.nnodes:>10} "
                "{limits.range.ncores:>10} "
                "{limits.range.ngpus:>10}"
            ),
        },
        "nodes": {
            "description": "Show queue status plus all/up/allocated/free nodes",
            "format": (
                "?:{queuem:<8.8} "
                "{color_enabled}{enabled:>2}{color_off} "
                "{color_started}{started:>2}{color_off} "
                "{resources.all.nnodes:>6} "
                "{resources.up.nnodes:>10} "
                "{resources.down.nnodes:>10} "
                "{resources.allocated.nnodes:>11} "
                "{resources.free.nnodes:>10}"
            ),
        },
    }

    def __init__(self, subcommand):
        initial_dict = {}
        for key, value in self.builtin_formats.items():
            initial_dict[key] = {"formats": value}
        super().__init__(
            name="flux-queue", subcommand=subcommand, initial_dict=initial_dict
        )

    def validate(self, path, conf):
        """Validate a loaded flux-queue config file as dictionary"""

        for key, value in conf.items():
            if key in ["list"]:
                for key2, val2 in value.items():
                    if key2 == "formats":
                        self.validate_formats(path, val2)
                    else:
                        raise ValueError(f"{path}: invalid key {key}.{key2}")
            else:
                raise ValueError(f"{path}: invalid key {key}")


class QueueLimitsEffectiveRange:
    """
    QueueLimits wrapper which supports a effective range (configured minimum
    to effective maximum minimum of limit or all resources in queue).
    """

    def __init__(self, limits, resources):
        self.__limits = limits
        self.__resources = resources
        self.__effective_max = {}
        for item in ("nnodes", "ncores", "ngpus"):
            self.__effective_max[item] = min(
                getattr(limits.max, item), getattr(resources.all, item)
            )

    @property
    def nnodes(self):
        effective_max = self.__effective_max["nnodes"]
        return f"{self.__limits.min.nnodes}-{effective_max}"

    @property
    def ncores(self):
        effective_max = self.__effective_max["ncores"]
        return f"{self.__limits.min.ncores}-{effective_max}"

    @property
    def ngpus(self):
        effective_max = self.__effective_max["ngpus"]
        return f"{self.__limits.min.ngpus}-{effective_max}"


class QueueLimitsRange:
    """
    QueueLimits wrapper which supports a string range "{min}-{max}"
    """

    def __init__(self, limits, resources):
        self.effective = QueueLimitsEffectiveRange(limits, resources)
        for item in ("nnodes", "ncores", "ngpus"):
            minimum = getattr(limits.min, item)
            maximum = getattr(limits.max, item)
            setattr(self, item, f"{minimum}-{maximum}")


class QueueLimitsWrapper:
    def __init__(self, info):
        self.__limits = info.limits
        self.range = QueueLimitsRange(info.limits, info.resources)

    def __getattr__(self, attr):
        # Forward most attribute lookups to underlying QueueLimits instance
        return getattr(self.__limits, attr)


class QueueInfoWrapper:
    def __init__(self, queue_info):
        self.__qinfo = queue_info
        self.is_started = queue_info.started
        self.is_enabled = queue_info.enabled
        self.limits = QueueLimitsWrapper(queue_info)

    def __getattr__(self, attr):
        try:
            return getattr(self.__qinfo, attr)
        except (KeyError, AttributeError):
            raise AttributeError("invalid QueueInfo attribute '{}'".format(attr))

    @property
    def scheduling(self):
        return "started" if self.is_started else "stopped"

    @property
    def submission(self):
        return "enabled" if self.is_enabled else "disabled"

    @property
    def queue(self):
        return self.name

    @property
    def queuem(self):
        if self.name and self.is_default:
            return f"{self.name}*"
        return self.name

    @property
    def color_enabled(self):
        return "\033[01;32m" if self.is_enabled else "\033[01;31m"

    @property
    def color_off(self):
        return "\033[0;0m"

    @property
    def enabled(self):
        return AltField("✔", "y") if self.is_enabled else AltField("✗", "n")

    @property
    def color_started(self):
        return "\033[01;32m" if self.is_started else "\033[01;31m"

    @property
    def started(self):
        return AltField("✔", "y") if self.is_started else AltField("✗", "n")


def generate_resource_headings():
    headings = {}
    for item in ("nnodes", "ncores", "ngpus"):
        for state in ("all", "free", "allocated", "up", "down"):
            heading = item[1:].upper()
            if state != "all":
                heading += f":{state[:5].upper()}"
            headings[f"resources.{state}.{item}"] = heading
    return headings


def list(args):
    headings = {
        "queue": "QUEUE",
        "queuem": "QUEUE",
        "submission": "SUBMIT",
        "scheduling": "SCHED",
        "enabled": "EN",
        "started": "ST",
        "enabled.ascii": "EN",
        "started.ascii": "ST",
        "color_enabled": "",
        "color_started": "",
        "color_off": "",
        "defaults.timelimit": "TDEFAULT",
        "limits.timelimit": "TLIMIT",
        "limits.range.nnodes": "NNODES",
        "limits.range.ncores": "NCORES",
        "limits.range.ngpus": "NGPUS",
        "limits.range.effective.nnodes": "NNODES",
        "limits.range.effective.ncores": "NCORES",
        "limits.range.effective.ngpus": "NGPUS",
        "limits.min.nnodes": "MINNODES",
        "limits.max.nnodes": "MAXNODES",
        "limits.min.ncores": "MINCORES",
        "limits.max.ncores": "MAXCORES",
        "limits.min.ngpus": "MINGPUS",
        "limits.max.ngpus": "MAXGPUS",
    }
    headings.update(generate_resource_headings())

    fmt = FluxQueueConfig("list").load().get_format_string(args.format)
    formatter = flux.util.OutputFormat(fmt, headings=headings)

    queues = [QueueInfoWrapper(x) for x in QueueList(flux.Flux(), queues=args.queue)]
    formatter.print_items(queues, no_header=args.no_header)


def queue_enable(handle, enable, name, all, reason=None):
    payload = {"enable": enable, "all": all}
    if name:
        payload["name"] = name
    if reason:
        payload["reason"] = reason
    future = handle.rpc("job-manager.queue-enable", payload, nodeid=0)
    try:
        future.get()
    except EnvironmentError:
        LOGGER.error("enable: {0}".format(future.error_string()))
        sys.exit(1)


def enable(args):
    handle = flux.Flux()
    queue_enable(handle, True, args.queue, args.all)
    if not args.quiet:
        queue_status(handle, args.queue, args.verbose, print_enable_status)


def disable(args):
    reason = "disabled by administrator"
    if args.message:
        reason = args.message
    handle = flux.Flux()
    queue_enable(handle, False, args.queue, args.all, reason)
    if not args.quiet:
        queue_status(handle, args.queue, args.verbose, print_enable_status)


def check_all(handle, name, all_queues):
    if not name and not all_queues:
        qlist = handle.rpc("job-manager.queue-list").get()
        if len(qlist["queues"]) > 0:
            raise ValueError("Named queues are defined. Specify queues or use --all.")


def queue_start(handle, start, name, all, nocheckpoint=False, reason=None):
    check_all(handle, name, all)
    payload = {"start": start, "all": all, "nocheckpoint": nocheckpoint}
    if name:
        payload["name"] = name
    if reason:
        payload["reason"] = reason
    future = handle.rpc("job-manager.queue-start", payload, nodeid=0)
    try:
        future.get()
    except EnvironmentError:
        LOGGER.error("start: {0}".format(future.error_string()))
        sys.exit(1)


def start(args):
    handle = flux.Flux()
    queue_start(handle, True, args.queue, args.all)
    if not args.quiet:
        queue_status(handle, args.queue, args.verbose, print_start_status)
        if args.verbose:
            alloc_query(handle)


def stop(args):
    reason = args.message
    handle = flux.Flux()
    queue_start(handle, False, args.queue, args.all, args.nocheckpoint, reason)
    if not args.quiet:
        queue_status(handle, args.queue, args.verbose, print_start_status)
        if args.verbose:
            alloc_query(handle)


def drain(args):
    handle = flux.Flux()
    future = handle.rpc("job-manager.drain")
    try:
        future.wait_for(args.timeout)
        future.get()
    except EnvironmentError as e:
        if e.errno == errno.ETIMEDOUT:
            LOGGER.error("drain: timeout")
        else:
            LOGGER.error("drain: {0}".format(future.error_string()))
        sys.exit(1)


def idle(args):
    handle = flux.Flux()
    future = handle.rpc("job-manager.idle")
    try:
        future.wait_for(args.timeout)
        rsp = future.get()
        pending = rsp["pending"]
    except EnvironmentError as e:
        if e.errno == errno.ETIMEDOUT:
            LOGGER.error("idle: timeout")
        else:
            LOGGER.error("idle: {0}".format(future.error_string()))
        sys.exit(1)
    if not args.quiet or pending > 0:
        print(f"{pending} jobs")


class FSDAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        time = parse_fsd(values)
        setattr(namespace, self.dest, time)


LOGGER = logging.getLogger("flux-queue")


def common_parser_create(subparsers, command):
    """
    Create an options parser for one of the flux-queue subcommands that
    take a common set of arguments. (status, enable, disable, start, stop)
    """
    command_parser = subparsers.add_parser(
        command, formatter_class=flux.util.help_formatter()
    )
    command_parser.add_argument(
        "-q",
        "--queue",
        dest="old_queue",
        type=str,
        metavar="NAME",
        help=argparse.SUPPRESS,
    )
    if command in ("enable", "disable", "start", "stop"):
        command_parser.add_argument(
            "-a",
            "--all",
            action="store_true",
            help="Force command to apply to all queues if none specified",
        )
        command_parser.add_argument(
            "--quiet", action="store_true", help="Display only errors"
        )
    command_parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Display more detail about internal job manager state",
    )
    if command in ("stop", "disable"):
        verb = dict(stop="stopped", disable="disabled")
        command_parser.add_argument(
            "--nocheckpoint",
            action="store_true",
            help=f"Do not checkpoint that the queue has been {verb[command]}",
        )
        command_parser.add_argument(
            "-m",
            "--message",
            type=str,
            metavar="REASON",
            help=f"Add reason that a queue is {verb[command]}",
        )
    command_parser.add_argument(
        "queue", metavar="QUEUE", nargs="?", help=f"Queue to {command}"
    )
    command_parser.set_defaults(func=globals()[command])
    return command_parser


@flux.util.CLIMain(LOGGER)
def main():
    sys.stdout = open(
        sys.stdout.fileno(), "w", encoding="utf8", errors="surrogateescape"
    )

    parser = argparse.ArgumentParser(prog="flux-queue")
    subparsers = parser.add_subparsers(
        title="subcommands", description="", dest="subcommand"
    )
    subparsers.required = True

    list_parser = subparsers.add_parser(
        "list", formatter_class=flux.util.help_formatter()
    )
    list_parser.add_argument(
        "-q",
        "--queue",
        action=FilterActionSetUpdate,
        default=set(),
        metavar="QUEUE,...",
        help="Include only specified queues in output",
    )
    list_parser.add_argument(
        "-o",
        "--format",
        default="default",
        help="Specify output format using Python's string format syntax "
        + "or a defined format by name (use 'help' to get a list of names)",
    )
    list_parser.add_argument(
        "-n", "--no-header", action="store_true", help="Suppress header output"
    )
    list_parser.set_defaults(func=list)

    for command in ("status", "enable", "disable", "start", "stop"):
        common_parser_create(subparsers, command)

    drain_parser = subparsers.add_parser(
        "drain", formatter_class=flux.util.help_formatter()
    )
    drain_parser.add_argument(
        "-t",
        "--timeout",
        action=FSDAction,
        metavar="DURATION",
        default=-1.0,
        help="timeout after DURATION",
    )
    drain_parser.set_defaults(func=drain)

    idle_parser = subparsers.add_parser(
        "idle", formatter_class=flux.util.help_formatter()
    )
    idle_parser.add_argument(
        "-t",
        "--timeout",
        action=FSDAction,
        metavar="DURATION",
        default=-1.0,
        help="timeout after DURATION",
    )
    idle_parser.add_argument(
        "--quiet",
        action="store_true",
        help="Only display pending job count if nonzero",
    )
    idle_parser.set_defaults(func=idle)

    args = parser.parse_args()

    # Handle use of old `-q, --queue` option for commands that now support
    # the specification of queues in free arguments:
    if "old_queue" in args and args.old_queue is not None:
        if args.queue is not None:
            raise ValueError(
                "Conflicting queues specified as "
                + f"--queue={args.old_queue} and {args.queue}"
            )
        args.queue = args.old_queue

    args.func(args)


if __name__ == "__main__":
    main()

# vi: ts=4 sw=4 expandtab
